synapse/handlers/saml_handler.py

```diff
@@ -205,9 +232,73 @@ def expire_sessions(self):
             del self._outstanding_requests_dict[reqid]
 
 
-@attr.s
-class Saml2SessionData:
-    """Data we track about SAML2 sessions"""
+DOT_REPLACE_PATTERN = re.compile(
+    ("[^%s]" % (re.escape("".join(mxid_localpart_allowed_characters)),))
+)
 
-    # time the session was created, in milliseconds
-    creation_time = attr.ib()
+
+def dot_replace_for_mxid(username: str) -> str:
+    username = username.lower()
+    username = DOT_REPLACE_PATTERN.sub(".", username)
+
+    # regular mxids aren't allowed to start with an underscore either
+    username = re.sub("^_", "", username)
+    return username
+
+
+MXID_MAPPER_MAP = {
+    "hexencode": map_username_to_mxid_localpart,
+    "dotreplace": dot_replace_for_mxid,
+}
+
+
+class DefaultSamlMappingProvider(object):
+    __version__ = "0.0.1"
+
+    def saml_response_to_user_attributes(
+        self,
+        config: dict,
+        saml_response: saml2.response.AuthnResponse,
+        failures: int = 0,
+    ) -> dict:
+        """Maps some text from a SAML response to attributes of a new user
+
+        Args:
+            config: A configuration dictionary
+
+            saml_response: A SAML auth response object
+
+            failures: How many times a call to this function with this
+                saml_response has resulted in a failure
+
+        Returns:
+            dict: A dict containing new user attributes. Possible keys:
+                * mxid_localpart (str): Required. The localpart of the user's mxid
+                * displayname (str): The displayname of the user
+        """
+        try:
+            mxid_source = saml_response.ava[config["mxid_source_attribute"]][0]
+        except KeyError:
+            logger.warning(
+                "SAML2 response lacks a '%s' attestation",
+                config["mxid_source_attribute"],
+            )
+            raise SynapseError(
+                400, "%s not in SAML2 response" % (config["_mxid_source_attribute"],)
```

Comment:
we should aim to read config settings during construction, since (a) it will expose config errors sooner (b) it will stop you confusing `mxid_source_attribute` with `_mxid_source_attribute`